### DISCUSSION TITLE
自动释放池的前世今生 ---- 深入解析 autoreleasepool · /autoreleasepool

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#289] [Announcements]
https://draveness.me/autoreleasepool

**Comment by yhvh009** [comment]
你好，请问下这个保存autoreleasePoolPage的双向链表只有一个么？也就是所有线程的autoreleasePoolPage都保存在一个链表中，还是每个线程保存一个自己的链表？并且链表头也就是链表的入口位置是保存在哪里呢？谁来控制呢？

**Comment by crmo** [comment]
博主你好，首先感谢你写的objc系列的文章，读后真的很过瘾，干货满满，在你的指引下，我还把runtime源码下下来拜读。
关于pop时kill掉child的问题，我在阅读源码之后有一些想法想和你交流。

> 我到现在也不是很清楚为什么要根据当前页的不同状态 kill 掉不同 child 的页面。

我们先看看释放的函数`releaseUntil `，它在释放的时候其实会一直顺着`parent`往前释放，只到找到参数`stop`传入的地址，也就是说可能一次性释放好几个`page`。

```
// 代码有所删减
void releaseUntil(id *stop)
{
    while (this->next != stop) {
        AutoreleasePoolPage *page = hotPage();
        
        while (page->empty()) {
            page = page->parent;
            setHotPage(page);
        }
        
        id obj = *--page->next;
        memset((void*)page->next, SCRIBBLE, sizeof(*page->next));
        
        if (obj != POOL_BOUNDARY) {
            objc_release(obj);
        }
    }
    
    setHotPage(this);
}
```

然后我们来看看有疑问的代码

```
        // memory: delete empty children
        if (DebugPoolAllocation  &&  page->empty()) {  // 分支1
            // special case: delete everything during page-per-pool debugging
            AutoreleasePoolPage *parent = page->parent;
            page->kill();
            setHotPage(parent);
        } else if (DebugMissingPools  &&  page->empty()  &&  !page->parent) { // 分支2
            // special case: delete everything for pop(top) 
            // when debugging missing autorelease pools
            page->kill();
            setHotPage(nil);
        } 
        else if (page->child) { // 分支3
            // hysteresis: keep one empty child if page is more than half fully
            if (page->lessThanHalfFull()) {
                page->child->kill();
            }
            else if (page->child->child) {
                page->child->child->kill();
            }
        }
```

这块代码的作用是删除空的子节点，释放内存。pop之后三种情况：
1. 当前`page`为空，直接kill掉当前`page`，然后把`parent`设置为`hotpage`；
2. 当前`page`为空，而且没有`parent`，kill掉当前`page`，`hotpage`置为空；
3. 当前`page`不为空，但是有`child`，如果当前`page`的空间占用不到一半，释放`child`，如果当前`page`的空间占用超过一半，且`child`还有`child`，直接释放这个孙子辈的`page`。（对于第三步注释中的解释是：keep one empty child if page is more than half fully）

我们再看看`kill`的实现，可以发现他是会顺着`child`一直往后释放，保证释放节点的`child page`都被释放了。

```
void kill()
{
    AutoreleasePoolPage *page = this;
    while (page->child) page = page->child;
    
    AutoreleasePoolPage *deathptr;
    do {
        deathptr = page;
        page = page->parent;
        if (page) {
            page->unprotect();
            page->child = nil;
            page->protect();
        }
        delete deathptr;
    } while (deathptr != this);
}
```

到这里就可以得出结论了：
1. pop之后，所有`child page`肯定都为空了，且当前`page`一定是`hotPage`
2. 系统为了节约内存，判断，如果当前`page`空间使用少于一半，就释放掉所有的`child page`，如果当前`page`空间使用大于一半，就从孙子`page`开始释放，预留一个`child page`。

**Comment by cqj1558** [comment]
（我到现在也不是很清楚为什么要根据当前页的不同状态 kill 掉不同 child 的页面）
个人认为这是一个时间与空间的取舍
（在一些资料中看到iOS实现可变数组时底层会用到一种表结构，存储数据太多之后会严重影响性能，所以不会等到表存储满就会创建新的表链上之前的表；而autoreleasepool存入大量对象时，与nsmutablearray存入对象有极其相似的性能特性，怀疑两者底层使用相似的存储技术）。

kill掉child是为了节省内存；
但是当占用一半时，很快需要创建新的page，所以保留一个child，牺牲内存来提升性能。

**Comment by ChenZeBin** [comment]
有个疑惑是关于autorelease对象的释放？autorelease对象是在runloop一个循环结束后释放，那么如果在这个类中，我等待10分钟吧，这时候应该经历了很多runloop迭代了，按理说这个autorelease对象应该被释放了，但是我打印弱指针引用的autorelease对象，发现还没有释放

**Comment by zhaoyali-iOS** [comment]
> @yhvh009
> 你好，请问下这个保存autoreleasePoolPage的双向链表只有一个么？也就是所有线程的autoreleasePoolPage都保存在一个链表中，还是每个线程保存一个自己的链表？并且链表头也就是链表的入口位置是保存在哪里呢？谁来控制呢？

一个线程有自己单独autoreleasePool链表，也有可能没有链表。链表的hotPage存储在TLS中，因为链表是双向的，通过hotpage就可以找到表头和表尾，不需要再单独存储表头

**Comment by btxkenshin** [comment]
> @crmo
> 博主你好，首先感谢你写的objc系列的文章，读后真的很过瘾，干货满满，在你的指引下，我还把runtime源码下下来拜读。
> 关于pop时kill掉child的问题，我在阅读源码之后有一些想法想和你交流。
> 
> > 我到现在也不是很清楚为什么要根据当前页的不同状态 kill 掉不同 child 的页面。
> 
> 我们先看看释放的函数`releaseUntil `，它在释放的时候其实会一直顺着`parent`往前释放，只到找到参数`stop`传入的地址，也就是说可能一次性释放好几个`page`。
> 
> ```
> // 代码有所删减
> void releaseUntil(id *stop)
> {
>     while (this->next != stop) {
>         AutoreleasePoolPage *page = hotPage();
>         
>         while (page->empty()) {
>             page = page->parent;
>             setHotPage(page);
>         }
>         
>         id obj = *--page->next;
>         memset((void*)page->next, SCRIBBLE, sizeof(*page->next));
>         
>         if (obj != POOL_BOUNDARY) {
>             objc_release(obj);
>         }
>     }
>     
>     setHotPage(this);
> }
> ```
> 
> 然后我们来看看有疑问的代码
> 
> ```
>         // memory: delete empty children
>         if (DebugPoolAllocation  &&  page->empty()) {  // 分支1
>             // special case: delete everything during page-per-pool debugging
>             AutoreleasePoolPage *parent = page->parent;
>             page->kill();
>             setHotPage(parent);
>         } else if (DebugMissingPools  &&  page->empty()  &&  !page->parent) { // 分支2
>             // special case: delete everything for pop(top) 
>             // when debugging missing autorelease pools
>             page->kill();
>             setHotPage(nil);
>         } 
>         else if (page->child) { // 分支3
>             // hysteresis: keep one empty child if page is more than half fully
>             if (page->lessThanHalfFull()) {
>                 page->child->kill();
>             }
>             else if (page->child->child) {
>                 page->child->child->kill();
>             }
>         }
> ```
> 
> 这块代码的作用是删除空的子节点，释放内存。pop之后三种情况：
> 1. 当前`page`为空，直接kill掉当前`page`，然后把`parent`设置为`hotpage`；
> 2. 当前`page`为空，而且没有`parent`，kill掉当前`page`，`hotpage`置为空；
> 3. 当前`page`不为空，但是有`child`，如果当前`page`的空间占用不到一半，释放`child`，如果当前`page`的空间占用超过一半，且`child`还有`child`，直接释放这个孙子辈的`page`。（对于第三步注释中的解释是：keep one empty child if page is more than half fully）
> 
> 我们再看看`kill`的实现，可以发现他是会顺着`child`一直往后释放，保证释放节点的`child page`都被释放了。
> 
> ```
> void kill()
> {
>     AutoreleasePoolPage *page = this;
>     while (page->child) page = page->child;
>     
>     AutoreleasePoolPage *deathptr;
>     do {
>         deathptr = page;
>         page = page->parent;
>         if (page) {
>             page->unprotect();
>             page->child = nil;
>             page->protect();
>         }
>         delete deathptr;
>     } while (deathptr != this);
> }
> ```
> 
> 到这里就可以得出结论了：
> 1. pop之后，所有`child page`肯定都为空了，且当前`page`一定是`hotPage`
> 2. 系统为了节约内存，判断，如果当前`page`空间使用少于一半，就释放掉所有的`child page`，如果当前`page`空间使用大于一半，就从孙子`page`开始释放，预留一个`child page`。

补充下：

1. releaseUntil 做的工作是真正擦除AutoreleasePoolPage所占用的内存区域的内容的（通过memset），但是并没有置空child或者parent指针

2. kill做的工作是置空已经被releaseUntil是放过的page的child的，也就是释放child所指向的AutoreleasePoolPage占用空间

   1. ```
      if (page->lessThanHalfFull()) {
          page->child->kill();
      }
      ```

      当前page的空间占用少于一半，释放掉child page占用的空间

   2. ```
      else if (page->child->child) {
          page->child->child->kill();
      }
      ```

      当前page的空间占用超过一半的，因为很可能会满了，就不释放child page的占用空间了，为了预留一个child page，满了的话我们就不用创建一个page再置为自己的child了，但是只要预留一页就够了，所以释放掉child的child，也就是只保留page和page+1页
